instrumentation/jms/src/main/java/brave/jms/MessageParser.java

```diff
@@ -36,16 +36,35 @@
   }
 
   @Nullable static String channelKind(@Nullable Destination destination) {
-    if (destination instanceof Queue) return "queue";
-    if (destination instanceof Topic) return "topic";
-    return null;
+    if (destination == null) return null;
+    return isQueue(destination) ? "queue" : "topic";
+  }
+
+  /**
+   * Handles special case of a destination being both a Queue and a Topic by checking if the
+   * {@linkplain Queue#getQueueName() queue name} is readable.
```

Comment:
Perhaps need to link to the issue here or provide a bit more background. The JMS APIs are generally sketchy, but the javadoc doesn't seem to indicate this name would ever return null.

Reply:
we check null anyway, especially as everything can throw JMSException and we coerce those to null. Our bug was one place where we didn't: this line is more about fallback behavior than null.

This era of library design didn't have clarifying annotations or docs on null as a matter of course. This matter is compounded by TCKs not being open source either.. I don't think it would help to assume non-null on JMS properties with or without this change.

I could add a comment about the example of something implementing both, but I'd hesitate to use this as a why something can return null as generally they can become null and don't want people mistakenly thinking null is an edge case.

wdyt?

Reply:
https://github.com/openzipkin/brave/pull/1100 hoping this topic doesn't eat our afternoon